src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/AttributeUpgradeAnalyzer.cs

```diff
@@ -0,0 +1,127 @@
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.AttributeUpgradeMessageFormat), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.AttributeUpgradeDescription), Resources.ResourceManager, typeof(Resources));
+
+        private static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private static void AnalyzeCSharpAttribute(SyntaxNodeAnalysisContext context, IEnumerable<TypeMapping> mappings) =>
+            AnalyzeAttribute(context, mappings, ((CSSyntax.AttributeSyntax)context.Node).Name.ToString());
+
+        private static void AnalyzeVBAttribute(SyntaxNodeAnalysisContext context, IEnumerable<TypeMapping> mappings) =>
+            AnalyzeAttribute(context, mappings, ((VBSyntax.AttributeSyntax)context.Node).Name.ToString());
+
+        private static void AnalyzeAttribute(SyntaxNodeAnalysisContext context, IEnumerable<TypeMapping> mappings, string? attributeName)
+        {
+            if (attributeName is null)
+            {
+                return;
+            }
+
+            // If the attribute name isn't one of the mapped names, bail out
+            var mapping = mappings.FirstOrDefault(m => MatchesPartiallyQualifiedName(m.OldName, attributeName) || MatchesPartiallyQualifiedName(m.OldName, $"{attributeName}{AttributeSuffix}"));
+            if (mapping is null)
+            {
+                return;
+            }
+
+            // Attempt to get the symbol for the attribute's type
+            var attrType = context.SemanticModel.GetTypeInfo(context.Node);
+
+            // Bail out if the node corresponds to a symbol that isn't the old type.
+            if (attrType.Type is INamedTypeSymbol typeSymbol
+                && typeSymbol is not IErrorTypeSymbol
+                && !typeSymbol.ToDisplayString(NullableFlowState.NotNull).Equals(mapping.OldName, StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            // Store the new attribute type that this identifier should be replaced with (or null to remove the attribute)
+            // for use by the code fix provider.
+            var properties = ImmutableDictionary.Create<string, string?>().Add(NewTypeKey, mapping.NewName);
+
+            // Create and report the diagnostic
+            var diagnostic = Diagnostic.Create(Rule, context.Node.GetLocation(), properties, mapping.OldName, mapping.NewName);
+            context.ReportDiagnostic(diagnostic);
+        }
+
+        private static bool MatchesPartiallyQualifiedName(string fullyQualifiedName, string partiallyQualifiedName)
```

Comment:
This would be a great addition for the NameMatcher class

Reply:
Good point. I've moved it there.